### DISCUSSION TITLE
Memory impact of RES

### DISCUSSION THREAD

**Discussion by pmackay** [RailsEventStore/rails_event_store#1311] [Q&A]
Hello, I'm interested to test RES. After adding the gem, I ran derailed benchmark and saw that it came just under rails itself:

`rails_event_store: 20.0469 MiB`

This seems high but I wonder how many of the dependencies are loaded by rails anyway?

Has there been any exploration into the impact of RES on memory usage? Also #1161 is an idea that could help to reduce this (if the browser isnt needed).

**Comment by mostlyobvious** [comment]
Can you please post the whole trace of this related to `rails_event_store` or a way to reproduce this result?

**Comment by mostlyobvious** [comment]
Hey @pmackay, I've [recreated the test](https://github.com/pawelpacana/derailed_res) but it does not match your findings. Perhaps we have a different test methodology, checked completely different version of RES/Ruby/Rails or interpret the results differently. I'd still love to know what were your conditions (versions, etc.) and how you got those results as it sounds a bit high to me too.

Anyway here's the result I got:

```
rails_event_store: 6.6406 MiB
    rails_event_store_active_record: 5.2656 MiB
      /Users/mostlyobvious/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/gems/rails_event_store_active_record-2.4.1/lib/rails_event_store_active_record/event: 5.2188 MiB
        active_record/base: 4.2656 MiB
          active_record/validations: 0.7188 MiB
            active_model/validations: 0.5625 MiB
          active_record/encryption: 0.5 MiB
          active_record/enum: 0.4063 MiB
            active_record/type: 0.3594 MiB
          active_record/querying: 0.3281 MiB
        active_record/relation: 0.4688 MiB
        active_record/associations/collection_proxy: 0.4688 MiB
    /Users/mostlyobvious/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/gems/rails_event_store-2.4.1/lib/rails_event_store/all: 1.375 MiB
      /Users/mostlyobvious/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/gems/rails_event_store-2.4.1/lib/rails_event_store/browser: 0.9844 MiB
        ruby_event_store/browser/app: 0.9844 MiB
          sinatra/base: 0.8594 MiB
```     

Most of it comes from ActiveRecord, which is unavoidable and already present in Rails application.

**Comment by mostlyobvious** [comment]
Regarding individual RES gem dependencies as of version 2.4.1:

```
ruby_event_store-2.4.1
  concurrent-ruby (~> 1.0, >= 1.1.6)
  
rails_event_store_active_record-2.4.1
  activerecord (>= 6.0)
  ruby_event_store (= 2.4.1)
  
ruby_event_store-browser-2.4.1
  ruby_event_store (= 2.4.1)
  sinatra (>= 0)
  
aggregate_root-2.4.1
  ruby_event_store (= 2.4.1)  
 
arkency-command_bus-0.4.1
  concurrent-ruby (>= 0)
```

The `rails_event_store` is currently an umbrella for the complete distribution, plus a few Rails-specific additions:

```
rails_event_store-2.4.1
  activejob (>= 3.0)
  activemodel (>= 3.0)
  activesupport (>= 3.0)
  aggregate_root (= 2.4.1)
  arkency-command_bus (>= 0.4)
  rails_event_store_active_record (= 2.4.1)
  ruby_event_store (= 2.4.1)
  ruby_event_store-browser (= 2.4.1)
```

**Comment by mostlyobvious** [comment]
Before version 2.4.0 there we few additional dependencies:
- `activerecord-import` to support bulk insert on ActiveRecord < 6.0
- `ruby2_keywords` to support Ruby 2.6 keyword arguments compatibility (while developing to support Ruby 3.1 at the same time)
- outdated `thread_safe` dependency on command bus gem

These are no more. We do plan to get rid of `sinatra` dependency of `ruby_event_store-browser` too [in the future](https://github.com/RailsEventStore/rails_event_store/issues/958).

**Comment by mostlyobvious** [comment]
Here's another _derailed_ data point from commercial app:

```
  rails_event_store: 3.4063 MiB
    /Users/mostlyobvious/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/gems/rails_event_store-2.3.0/lib/rails_event_store/all: 2.1563 MiB
      /Users/mostlyobvious/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/gems/rails_event_store-2.3.0/lib/rails_event_store/browser: 1.8906 MiB
        ruby_event_store/browser/app: 1.8438 MiB
          sinatra/base: 1.0313 MiB
    rails_event_store_active_record: 1.25 MiB
      /Users/mostlyobvious/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/gems/rails_event_store_active_record-2.3.0/lib/rails_event_store_active_record/event_repository: 0.7344 MiB
        activerecord-import: 0.7031 MiB (Also required by: /Users/mostlyobvious/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/gems/rails_event_store_active_record-2.3.0/lib/rails_event_store_active_record/pg_linearized_event_repository)
          activerecord-import/base: 0.7031 MiB
            activerecord-import/import: 0.3438 MiB
            activerecord-import/active_record/adapters/abstract_adapter: 0.3438 MiB
              active_record/connection_adapters/abstract_adapter: 0.3281 MiB
      /Users/mostlyobvious/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/gems/rails_event_store_active_record-2.3.0/lib/rails_event_store_active_record/generators/migration_generator: 0.3906 MiB
```

**Comment by mostlyobvious** [comment]
Sinatra dependency is gone: https://github.com/RailsEventStore/rails_event_store/releases/tag/v2.5.0
